File: apps/frontend/src/features/accounts/schemas.ts
import { z } from "zod";

export const UsageTrendPointSchema = z.object({
  t: z.string().datetime({ offset: true }),
  v: z.number(),
});

export const AccountUsageTrendSchema = z.object({
  primary: z.array(UsageTrendPointSchema),
  secondary: z.array(UsageTrendPointSchema),
});

export const AccountUsageSchema = z.object({
  primaryRemainingPercent: z.number().nullable(),
  secondaryRemainingPercent: z.number().nullable(),
});

export const AccountRequestUsageSchema = z.object({
  requestCount: z.number().int().nonnegative(),
  totalTokens: z.number().int().nonnegative(),
  outputTokens: z.number().int().nonnegative().optional(),
  cachedInputTokens: z.number().int().nonnegative(),
  cacheWriteTokens: z.number().int().nonnegative().optional(),
  totalCostUsd: z.number().nonnegative(),
});

export const AccountTokenStatusSchema = z.object({
  expiresAt: z.string().datetime({ offset: true }).nullable().optional(),
  state: z.string().nullable().optional(),
});

export const AccountAuthSchema = z.object({
  access: AccountTokenStatusSchema.nullable().optional(),
  refresh: AccountTokenStatusSchema.nullable().optional(),
  idToken: AccountTokenStatusSchema.nullable().optional(),
});

export const AccountCodexAuthSchema = z.object({
  hasSnapshot: z.boolean(),
  snapshotName: z.string().nullable().optional(),
  activeSnapshotName: z.string().nullable().optional(),
  isActiveSnapshot: z.boolean().optional(),
  hasLiveSession: z.boolean().optional(),
  liveUsageConfidence: z.enum(["high", "low"]).nullable().optional(),
  expectedSnapshotName: z.string().nullable().optional(),
  snapshotNameMatchesEmail: z.boolean().optional(),
  runtimeReady: z.boolean().optional(),
  runtimeReadySource: z.enum(["validated_snapshot_email_match"]).nullable().optional(),
  isOmxBoosted: z.boolean().optional(),
});

export const AccountAdditionalWindowSchema = z.object({
  usedPercent: z.number(),
  resetAt: z.number().nullable().optional(),
  windowMinutes: z.number().nullable().optional(),
});

export const AccountAdditionalQuotaSchema = z.object({
  quotaKey: z.string().nullable().optional(),
  limitName: z.string(),
  meteredFeature: z.string(),
  displayLabel: z.string().nullable().optional(),
  primaryWindow: AccountAdditionalWindowSchema.nullable().optional(),
  secondaryWindow: AccountAdditionalWindowSchema.nullable().optional(),
});

export const AccountLiveQuotaDebugWindowSchema = z.object({
  usedPercent: z.number(),
  remainingPercent: z.number(),
  resetAt: z.number().nullable().optional(),
  windowMinutes: z.number().nullable().optional(),
});

export const AccountLiveQuotaDebugSampleSchema = z.object({
  source: z.string(),
  snapshotName: z.string().nullable().optional(),
  recordedAt: z.string().datetime({ offset: true }),
  stale: z.boolean().optional(),
  primary: AccountLiveQuotaDebugWindowSchema.nullable().optional(),
  secondary: AccountLiveQuotaDebugWindowSchema.nullable().optional(),
});

export const AccountLiveQuotaDebugSchema = z.object({
  snapshotsConsidered: z.array(z.string()).default([]),
  rawSamples: z.array(AccountLiveQuotaDebugSampleSchema).default([]),
  merged: AccountLiveQuotaDebugSampleSchema.nullable().optional(),
  overrideApplied: z.boolean().optional(),
  overrideReason: z.string().nullable().optional(),
});

export const AccountSessionTaskPreviewSchema = z.object({
  sessionKey: z.string(),
  taskPreview: z.string().nullable().optional().default(null),
  taskUpdatedAt: z.string().datetime({ offset: true }).nullable().optional().default(null),
});

export const AccountSummarySchema = z.object({
  accountId: z.string(),
  email: z.string(),
  displayName: z.string(),
  planType: z.string(),
  status: z.string(),
  deactivationReason: z.string().nullable().optional(),
  usage: AccountUsageSchema.nullable().optional(),
  resetAtPrimary: z.string().datetime({ offset: true }).nullable().optional(),
  resetAtSecondary: z.string().datetime({ offset: true }).nullable().optional(),
  lastUsageRecordedAtPrimary: z.string().datetime({ offset: true }).nullable().optional(),
  lastUsageRecordedAtSecondary: z.string().datetime({ offset: true }).nullable().optional(),
  windowMinutesPrimary: z.number().nullable().optional(),
  windowMinutesSecondary: z.number().nullable().optional(),
  requestUsage: AccountRequestUsageSchema.nullable().optional(),
  codexLiveSessionCount: z.number().int().nonnegative().optional(),
  codexTrackedSessionCount: z.number().int().nonnegative().optional(),
  codexSessionCount: z.number().int().nonnegative().optional(),
  codexCurrentTaskPreview: z.string().nullable().optional(),
  codexLastTaskPreview: z.string().nullable().optional(),
  codexSessionTaskPreviews: z.array(AccountSessionTaskPreviewSchema).optional(),
  liveQuotaDebug: AccountLiveQuotaDebugSchema.nullable().optional(),
  auth: AccountAuthSchema.nullable().optional(),
  codexAuth: AccountCodexAuthSchema.nullable().optional(),
  additionalQuotas: z.array(AccountAdditionalQuotaSchema).default([]),
});

export const AccountTrendsResponseSchema = z.object({
  accountId: z.string(),
  primary: z.array(UsageTrendPointSchema),
  secondary: z.array(UsageTrendPointSchema),
});

export const AccountsResponseSchema = z.object({
  accounts: z.array(AccountSummarySchema),
});

export const AccountImportResponseSchema = z.object({
  accountId: z.string(),
  email: z.string(),
  planType: z.string(),
  status: z.string(),
});

export const AccountActionResponseSchema = z.object({
  status: z.string(),
});

export const AccountUseLocalResponseSchema = z.object({
  status: z.string(),
  accountId: z.string(),
  snapshotName: z.string(),
});

export const AccountRefreshAuthResponseSchema = z.object({
  status: z.string(),
  accountId: z.string(),
  email: z.string(),
  planType: z.string(),
});

export const AccountOpenTerminalResponseSchema = z.object({
  status: z.string(),
  accountId: z.string(),
  snapshotName: z.string(),
});

export const AccountTerminateCliSessionsResponseSchema = z.object({
  status: z.string(),
  accountId: z.string(),
  snapshotName: z.string(),
  terminatedSessionCount: z.number().int().nonnegative(),
});

export const AccountSnapshotRepairResponseSchema = z.object({
  status: z.string(),
  accountId: z.string(),
  previousSnapshotName: z.string(),
  snapshotName: z.string(),
  mode: z.enum(["readd", "rename"]),
  changed: z.boolean(),
});

export const OauthStartRequestSchema = z.object({
  forceMethod: z.string().optional(),
});

export const OauthStartResponseSchema = z.object({
  method: z.string(),
  authorizationUrl: z.string().nullable(),
  callbackUrl: z.string().nullable(),
  verificationUrl: z.string().nullable(),
  userCode: z.string().nullable(),
  deviceAuthId: z.string().nullable(),
  intervalSeconds: z.number().nullable(),
  expiresInSeconds: z.number().nullable(),
});

export const OauthStatusResponseSchema = z.object({
  status: z.string(),
  errorMessage: z.string().nullable(),
});

export const OauthCompleteRequestSchema = z.object({
  deviceAuthId: z.string().optional(),
  userCode: z.string().optional(),
});

export const OauthCompleteResponseSchema = z.object({
  status: z.string(),
});

export const ManualOauthCallbackRequestSchema = z.object({
  callbackUrl: z.string(),
});

export const ManualOauthCallbackResponseSchema = z.object({
  status: z.string(),
  errorMessage: z.string().nullable(),
});

export const RuntimeConnectAddressResponseSchema = z.object({
  connectAddress: z.string(),
});

export const OAuthStateSchema = z.object({
  status: z.enum(["idle", "starting", "pending", "success", "error"]),
  method: z.enum(["browser", "device"]).nullable(),
  authorizationUrl: z.string().nullable(),
  callbackUrl: z.string().nullable(),
  verificationUrl: z.string().nullable(),
  userCode: z.string().nullable(),
  deviceAuthId: z.string().nullable(),
  intervalSeconds: z.number().nullable(),
  expiresInSeconds: z.number().nullable(),
  errorMessage: z.string().nullable(),
});

export const ImportStateSchema = z.object({
  status: z.enum(["idle", "uploading", "success", "error"]),
  message: z.string().nullable(),
});

export type UsageTrendPoint = z.infer<typeof UsageTrendPointSchema>;
export type AccountUsageTrend = z.infer<typeof AccountUsageTrendSchema>;
export type AccountSummary = z.infer<typeof AccountSummarySchema>;
export type AccountAdditionalWindow = z.infer<typeof AccountAdditionalWindowSchema>;
export type AccountAdditionalQuota = z.infer<typeof AccountAdditionalQuotaSchema>;
export type AccountLiveQuotaDebug = z.infer<typeof AccountLiveQuotaDebugSchema>;
export type AccountLiveQuotaDebugSample = z.infer<typeof AccountLiveQuotaDebugSampleSchema>;
export type AccountLiveQuotaDebugWindow = z.infer<typeof AccountLiveQuotaDebugWindowSchema>;
export type AccountTrendsResponse = z.infer<typeof AccountTrendsResponseSchema>;
export type AccountUseLocalResponse = z.infer<typeof AccountUseLocalResponseSchema>;
export type AccountRefreshAuthResponse = z.infer<typeof AccountRefreshAuthResponseSchema>;
export type AccountOpenTerminalResponse = z.infer<typeof AccountOpenTerminalResponseSchema>;
export type AccountTerminateCliSessionsResponse = z.infer<
  typeof AccountTerminateCliSessionsResponseSchema
>;
export type AccountSnapshotRepairResponse = z.infer<typeof AccountSnapshotRepairResponseSchema>;
export type OauthStartResponse = z.infer<typeof OauthStartResponseSchema>;
export type OauthStatusResponse = z.infer<typeof OauthStatusResponseSchema>;
export type ManualOauthCallbackResponse = z.infer<typeof ManualOauthCallbackResponseSchema>;
export type RuntimeConnectAddressResponse = z.infer<
  typeof RuntimeConnectAddressResponseSchema
>;
export type OAuthState = z.infer<typeof OAuthStateSchema>;
export type ImportState = z.infer<typeof ImportStateSchema>;

File: app/modules/accounts/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import List, Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class UsageTrendPoint(DashboardModel):
    t: datetime
    v: float


class AccountUsageTrend(DashboardModel):
    primary: list[UsageTrendPoint] = Field(default_factory=list)
    secondary: list[UsageTrendPoint] = Field(default_factory=list)


class AccountUsage(DashboardModel):
    primary_remaining_percent: float | None = None
    secondary_remaining_percent: float | None = None


class AccountRequestUsage(DashboardModel):
    request_count: int = 0
    total_tokens: int = 0
    output_tokens: int = 0
    cached_input_tokens: int = 0
    cache_write_tokens: int = 0
    total_cost_usd: float = 0.0


class AccountTokenStatus(DashboardModel):
    expires_at: datetime | None = None
    state: str | None = None


class AccountAuthStatus(DashboardModel):
    access: AccountTokenStatus | None = None
    refresh: AccountTokenStatus | None = None
    id_token: AccountTokenStatus | None = None


class AccountCodexAuthStatus(DashboardModel):
    has_snapshot: bool = False
    snapshot_name: str | None = None
    active_snapshot_name: str | None = None
    is_active_snapshot: bool = False
    has_live_session: bool = False
    live_usage_confidence: Literal["high", "low"] | None = None
    expected_snapshot_name: str | None = None
    snapshot_name_matches_email: bool = False
    runtime_ready: bool = False
    runtime_ready_source: Literal["validated_snapshot_email_match"] | None = None
    is_omx_boosted: bool = False
    runtime_mode: Literal["local", "cloud"] = "local"
    daemon_id: str | None = None
    device: str | None = None
    cli_version: str | None = None
    latest_cli_version: str | None = None
    cli_update_available: bool = False
    cli_update_command: str | None = None


class AccountAdditionalWindow(DashboardModel):
    used_percent: float
    reset_at: int | None = None
    window_minutes: int | None = None


class AccountAdditionalQuota(DashboardModel):
    quota_key: str | None = None
    limit_name: str
    metered_feature: str
    display_label: str | None = None
    primary_window: AccountAdditionalWindow | None = None
    secondary_window: AccountAdditionalWindow | None = None


class AccountLiveQuotaDebugWindow(DashboardModel):
    used_percent: float
    remaining_percent: float
    reset_at: int | None = None
    window_minutes: int | None = None


class AccountLiveQuotaDebugSample(DashboardModel):
    source: str
    snapshot_name: str | None = None
    recorded_at: datetime
    stale: bool = False
    primary: AccountLiveQuotaDebugWindow | None = None
    secondary: AccountLiveQuotaDebugWindow | None = None


class AccountLiveQuotaDebug(DashboardModel):
    snapshots_considered: list[str] = Field(default_factory=list)
    raw_samples: list[AccountLiveQuotaDebugSample] = Field(default_factory=list)
    merged: AccountLiveQuotaDebugSample | None = None
    override_applied: bool = False
    override_reason: str | None = None


class AccountSessionTaskPreview(DashboardModel):
    session_key: str
    task_preview: str | None = None
    task_updated_at: datetime | None = None


class AccountSummary(DashboardModel):
    account_id: str
    email: str
    display_name: str
    plan_type: str
    status: str
    usage: AccountUsage | None = None
    reset_at_primary: datetime | None = None
    reset_at_secondary: datetime | None = None
    last_usage_recorded_at_primary: datetime | None = None
    last_usage_recorded_at_secondary: datetime | None = None
    window_minutes_primary: int | None = None
    window_minutes_secondary: int | None = None
    last_refresh_at: datetime | None = None
    capacity_credits_primary: float | None = None
    remaining_credits_primary: float | None = None
    capacity_credits_secondary: float | None = None
    remaining_credits_secondary: float | None = None
    request_usage: AccountRequestUsage | None = None
    codex_live_session_count: int = 0
    codex_tracked_session_count: int = 0
    codex_session_count: int = 0
    codex_current_task_preview: str | None = None
    codex_last_task_preview: str | None = None
    codex_session_task_previews: list[AccountSessionTaskPreview] = Field(default_factory=list)
    live_quota_debug: AccountLiveQuotaDebug | None = None
    additional_quotas: list[AccountAdditionalQuota] = Field(default_factory=list)
    deactivation_reason: str | None = None
    auth: AccountAuthStatus | None = None
    codex_auth: AccountCodexAuthStatus | None = None


class AccountsResponse(DashboardModel):
    accounts: List[AccountSummary] = Field(default_factory=list)


class AccountImportResponse(DashboardModel):
    account_id: str
    email: str
    plan_type: str
    status: str


class AccountPauseResponse(DashboardModel):
    status: str


class AccountReactivateResponse(DashboardModel):
    status: str


class AccountDeleteResponse(DashboardModel):
    status: str


class AccountUseLocalResponse(DashboardModel):
    status: str
    account_id: str
    snapshot_name: str


class AccountRefreshAuthResponse(DashboardModel):
    status: str
    account_id: str
    email: str
    plan_type: str


class AccountOpenTerminalResponse(DashboardModel):
    status: str
    account_id: str
    snapshot_name: str


class AccountTerminateCliSessionsResponse(DashboardModel):
    status: str
    account_id: str
    snapshot_name: str
    terminated_session_count: int


class AccountSnapshotRepairResponse(DashboardModel):
    status: str
    account_id: str
    previous_snapshot_name: str
    snapshot_name: str
    mode: str
    changed: bool


class AccountTrendsResponse(DashboardModel):
    account_id: str
    primary: list[UsageTrendPoint] = Field(default_factory=list)
    secondary: list[UsageTrendPoint] = Field(default_factory=list)
